### PR TITLE
Encode render passes while command buffer is only read-locked

### LINKED ISSUE #1104: Allow Parallel Recording of RenderPasses
**Is your feature request related to a problem? Please describe.**

Currently, the best way to multithreaded the command recording of wgpu applications is to spread the job of recording onto multiple threads. However, due to a singular write lock we take on the command buffer list [here](https://github.com/gfx-rs/wgpu/blob/master/wgpu-core/src/command/render.rs#L522), it enforces that only one RenderPass will be running at once.

**Describe the solution you'd like**

Demote this write lock to a read lock. The simplest way is to make the registry hold `Registry<RwLock<CommandBuffer<B>>, CommandBufferId, F>`. There may be better ideas.

**Describe alternatives you've considered**

None. This is a pretty vital feature and one of the big appeals of modern apis, so we want to make sure this works well.

**Additional context**

This came up when someone was trying to improve encoding time of a program with tens of thousands of direct draw calls. Parallelizing the job made it perform no better.

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
Could fix #1104 to some extent

**Description**
The heaviest part of the render pass encoding doesn't need a write lock. Only in the end we need it.

**Testing**
Untested.

### CODE REVIEW

**Comment by kvark** [comment]
Note: this would benefit from a good refactor to reduce nesting. I don't want to invest in the refactor until we know if this is the way to go.

**Comment by kvark** [comment]
@cwfitzgerald I took a stab at making the code nicer in the last commit. Please check it out.
The PR should be ready for landing otherwise.

File: wgpu-core/src/command/mod.rs [hunk 1]
```diff
@@ -64,6 +64,17 @@ impl<B: GfxBackend> CommandBuffer<B> {
         }
     }
 
+    fn get_encoder_read_only(
```
Comment (cwfitzgerald) [inline wgpu-core/src/command/mod.rs:67]: Nit: I think it would be better to have this follow thee `get_encoder` `get_encoder_mut` pattern

**Review by cwfitzgerald — APPROVED** [review]
Beautiful, thank you for tackling this one.

One very minor nit which you can take or leave.

**Comment by kvark** [comment]
bors r=cwfitzgerald

**Pull request merged**